### PR TITLE
Add a white list to collect timeline events on a filtered events

### PR DESCRIPTION
**Pull request by huaqiangwang**
By default, if the TIMELINE is enabled the timeline tracing output json file size will be huge if running for several minutes. This PR is addressing this problem to pass a timeline event filter through 'XFT_TIMELINE_WHITELIST' os environment.

For example, if you only care about the first and second token events, you could set the environment to ignore all other timeline events:

export XFT_TIMELINE_WHITELIST='1st Token,Next Token'

This PR also introduces the following changes:
- test: add timeline unit test
- cmake: add the WITH_TIMELINE knob for timeline event
      to enable TIMELINE: cmake -DWITH_TIMELINE=ON .
- empty timeline pool after dump_pool()

Br

### CODE REVIEW

**Comment by pujiang2018** [comment]
@changqi1 Could you please review this PR?

File: src/utils/timeline.h [hunk 1]
```diff
@@ -171,23 +174,77 @@ class TimeLine {
...
     std::string tag_name;
     int64_t pid;
     int64_t tid;
     std::chrono::high_resolution_clock::time_point start;
     std::chrono::high_resolution_clock::time_point end;
-    int64_t start_timestap, during_time;
+    int64_t start_timestamp, during_time;
     Json::Value trace_event;
+    static inline std::vector<Json::Value> pool{};
+    static inline std::unordered_map<std::string, int32_t> tag_whitelist{};
```
Comment (changqi1) [inline src/utils/timeline.h:235]: Sorry, I don't know why to use std::unordered_map structure. vector?

**Comment by changqi1** [comment]
Hump naming and format code.

File: src/utils/timeline.h (hunk 1 above)
Comment (changqi1) [reply src/utils/timeline.h:235]: Got. for avoiding duplicate lookups

Comment (huaqiangwang) [reply src/utils/timeline.h:235]: This is for the case of a large 'whitelist', looking for an event name in unordered_map will be faster than vector list. A unordered_map is a "hash_map", which means that searching it is O(1).

File: src/models/models.cpp [hunk 2]
```diff
@@ -187,6 +188,7 @@ void Model::unsetPrefix() {
 AutoModel::AutoModel(std::string modelPath, xft::DataType datatype) : Model() {
     std::string configPath = modelPath + "/config.ini";
     INIReader reader = INIReader(configPath);
+    TimeLine::init();
```
[the remark this answers is not in the document]
Comment (huaqiangwang) [reply src/models/models.cpp:191]: Sure. Will move to that with a following commit.

**Review by changqi1 — APPROVED** [review]

**Pull request merged**